errors/middleware-upgrade-guide.md

```diff
@@ -0,0 +1,290 @@
+# Middleware Upgrade Guide
+
+As we work on improving Middleware for General Availability (GA), we've made some changes to the Middleware APIs (and how you define Middleware in your application) based on your feedback.
+
+This upgrade guide will help you understand the changes and how to migrate your existing Middleware to the new API. This guide is for Next.js customers who:
+
+- Currently use the beta Next.js Middleware features
+- Choose to upgrade to the next stable version of Next.js
+
+> **Note:** For customers using Next.js on Vercel, your existing deploys using Middlware will continue to work. Existing sites can be redeployed without needing to upgrade. Before upgrading to the next stable version of Next.js, you will need to follow the upgrade guide below.
+
+## Breaking changes
+
+1. [No Nested Middleware](#no-nested-middleware)
+2. [No Response Body](#no-response-body)
+3. [Cookies API Revamped](#cookies-api-revamped)
+4. [No More Page Match Data](#no-more-page-match-data)
+5. [Executing Middleware on Internal Next.js Requests](#executing-middleware-on-internal-next.js-requests)
+
+## No Nested Middleware
+
+### Summary of changes
+
+- Define a single Middleware file at the root of your project
+- No need to prefix the file with an underscore
+- A custom matcher can be used to define matching routes using an exported config object
+
+### Explanation
+
+Previously, you could create a `_middleware.js` file under the `pages` directory at any level. Middleware execution was based on the file path where it was created. Beta customers found this route matching confusing. For example:
+
+- Middleware in `pages/dashboard/_middleware.ts`
+- Middlware in `pages/dashboard/users/_middleware.ts`
+- A request to `/dashboard/users/*` **would match both.**
+
+Based on customer feedback, we have replaced this API with a single root Middleware.
+
+### How to upgrade
+
+You should declare **one single Middleware file** in your application, which should be located at the root of the project directory (**not** inside of the `pages` directory) and named without an `_` prefix. Your Middleware file can still have either a `.ts` or `.js` extension.
+
+Middleware will be invoked for **every route in the app**, and a custom matcher can be used to define matching filters. The following is an example for a Middleware that triggers for `/about/*`, the custom matcher is defined in an exported config object:
+
+```typescript
+// middleware.ts
+import type { NextRequest } from 'next/server'
+
+export function middleware(request: NextRequest) {
+  return NextResponse.rewrite(new URL('/about-2', request.url))
+}
+// config with custom matcher
+export const config = {
+  pathname: '/about/:path*',
+}
+```
+
+You can also use a conditional statement to only run the Middleware when it matches the `about/*` path:
```

Comment:
I guess the difference between the two approaches is that the middleware in the approach below gets invoked much more often (for every request, vs only for requests to `/about/*`)?

Might be nice to call that difference out explicitly

Reply:
Updated